### PR TITLE
Introduce Decimal Comma number handling.

### LINKED ISSUE #100: Convert to float should handle european string values "9,99" better
from AMA in London elasticon 2017

### PR DESCRIPTION
**Pull request by guyboertje**
Support number strings using a **decimal comma** (e.g. 1,23), added convert support to specify integer_eu and float_eu

Fixes #100

### CODE REVIEW

File: lib/logstash/filters/mutate.rb [hunk 1]
```diff
@@ -300,7 +304,7 @@ def convert(event)
       when Hash
         @logger.debug? && @logger.debug("I don't know how to type convert a hash, skipping", :field => field, :value => original)
       when Array
-        event.set(field, original.map { |v| converter.call(v) })
+        event.set(field, original.map { |v| v.nil? ? v : converter.call(v) })
```
Comment (colinsurprenant) [inline lib/logstash/filters/mutate.rb]: does this fix a known issue?

File: lib/logstash/filters/mutate.rb [hunk 2]
```diff
@@ -320,21 +324,32 @@ def convert_string(value)
...
+    return false if value.empty? || value =~ FALSE_REGEX
+    @logger.warn("Failed to convert #{value} into boolean.")
+    value
+  end
+
   def convert_integer(value)
     return 1 if value == true
     return 0 if value == false
-    value.to_i
+    value.squeeze(MUC_CP).tr(MUC_CP, MUC_UP).to_i
```
Comment (colinsurprenant) [inline lib/logstash/filters/mutate.rb]: I am not sure about the purpose of `squeeze` here? could you please explain?

File: lib/logstash/filters/mutate.rb [hunk 3]
```diff
@@ -320,21 +324,32 @@ def convert_string(value)
...
 
-  def convert_boolean(value)
-    return true if value =~ TRUE_REGEX
-    return false if value.empty? || value =~ FALSE_REGEX
-    @logger.warn("Failed to convert #{value} into boolean.")
-    value
+  def convert_integer_eu(value)
+    return 1 if value == true
+    return 0 if value == false
+    value.squeeze(MUC_CP).tr(MUC_CP.reverse, MUC_UP).to_i
```
Comment (colinsurprenant) [inline lib/logstash/filters/mutate.rb]: is converting to `0` and `1` from `false` and `true` a documented/historical  behaviour or a new feature introduced here?

File: lib/logstash/filters/mutate.rb [hunk 4]
```diff
@@ -320,21 +324,32 @@ def convert_string(value)
...
   end
 
-  def convert_boolean(value)
-    return true if value =~ TRUE_REGEX
-    return false if value.empty? || value =~ FALSE_REGEX
-    @logger.warn("Failed to convert #{value} into boolean.")
-    value
+  def convert_integer_eu(value)
+    return 1 if value == true
+    return 0 if value == false
```
Comment (colinsurprenant) [inline lib/logstash/filters/mutate.rb]: is converting to 0 and 1 from false and true a documented/historical behaviour or a new feature introduced here?

Comment (colinsurprenant) [reply lib/logstash/filters/mutate.rb]: nvmd - I see that `convert_integer` was already doing that.

**Comment by colinsurprenant** [comment]
As per our discussion, I am wondering why we have to squeeze and translate grouping to underscore "_"?

Why not just remove the group separator? (and set decimal separator to `.` if needed)  For example:
-  integer_eu or float_eu: `"1..234.567,89".tr(".","").tr(",", ".")`
```
"1..234.567,89".tr(".","").tr(",", ".").to_f
=> 1234567.89
```
- integer or float: `"1,,234,567.89".tr(",","")`
```
"1,,234,567.89".tr(",","").to_f
=> 1234567.89
```

**Comment by jordansissel** [comment]
What about having a Locale setting instead of `integer_eu` ? Would locale make sense to have instead of an 'integer_eu' style setting?

File: lib/logstash/filters/mutate.rb (hunk 4 above)
Comment (jordansissel) [reply lib/logstash/filters/mutate.rb]: Thoughts on using [java.text.NumberFormat](https://docs.oracle.com/javase/8/docs/api/java/text/NumberFormat.html).[parse](https://docs.oracle.com/javase/8/docs/api/java/text/NumberFormat.html)() which supports Locales and whatnot?

Comment (guyboertje) [reply lib/logstash/filters/mutate.rb]: its as above for `convert_integer`

Comment (guyboertje) [reply lib/logstash/filters/mutate.rb]: I checked to see if JRuby used the Java text NumberFormatter but it does not. I presume to follow what CRuby does. I am concerned that NumberFormat **may** introduce new behaviour after reading this https://www.ibm.com/developerworks/library/j-numberformat/

**Comment by guyboertje** [comment]
I considered using Locale for both a NumberFormat based solution and a Ruby solution.

JRuby is not locale aware when converting strings to numbers - it hard codes the `_` and `.` parsing.

Obviously we could build a NumberFormat instance that is Locale aware (with the caveat that it might behave differently, as in, why didn't JRuby use NumberFormat?). These concerns become more apparent when we consider moving the internal of certain plugins to Java.

I struggled to think of a way to build a setting that allowed a user to specify which fields to apply certain fields to. I considered a centralised logging case where events have some fields in one format and other fields in the other format. One Locale per filter could force the user to declare two filters in their configs. A separate setting, say `number_formats => { "fr-CA" => ["[field-a]", "[field-b]"] "de-DE" => ["[field-c]", "[field-d]"]`, would require referring to `field-a` twice.

**Comment by colinsurprenant** [comment]
I do not have a preference for using a Locale setting or not. I believe there are only 2 types of Locales that applies here? 

@guyboertje and what about my suggestion for above to avoid using `_` in the conversion?

File: lib/logstash/filters/mutate.rb (hunk 2 above)
Comment (guyboertje) [reply lib/logstash/filters/mutate.rb]: As discussed on AON, it is to replace multiple consecutive commas and periods. But I take your point that this should probably flag up anomaly between `to_i` and `to_f` handling of multiple `_` separators.

File: lib/logstash/filters/mutate.rb [hunk 5]
```diff
@@ -320,21 +320,37 @@ def convert_string(value)
...
+    @logger.warn("Failed to convert #{value} into boolean.")
+    value
+  end
+
   def convert_integer(value)
     return 1 if value == true
     return 0 if value == false
-    value.to_i
+    return value.to_i if !value.is_a?(String)
+    value.tr(",", "").to_i
```
Comment (colinsurprenant) [inline lib/logstash/filters/mutate.rb]: minor/style: maybe `return value.is_a?(String) ? value.tr(",", "").to_i : value.to_i` ?

**Comment by colinsurprenant** [comment]
Left a minor style note, feel free to ignore.
It seems we are missing the docs for `integer_eu` and `float_eu` in the config section comments?

**Comment by colinsurprenant** [comment]
@guyboertje I am personally ok with `integer_eu` and `float_eu` converters notation unless @jordansissel has objections, and we're good with my above comment on the inline docs, I'll be good for a LGTM.

**Comment by jordansissel** [comment]
I'm OK with the proposal as is.

My thought for Locale came from my belief of the intent of the 'convert' feature - The intent with convert, at least as I see it, is for taking a string "123456" and making it a numeric value (in memory, json). It was not intended to convert human-written numeric styled text "1,234,567" or Ruby "1_234_567" or EU-style "1.234.567' etc -- and if it ever did/does then maybe that's an accident. We can embrace this as a feature, as this PR does.

**Comment by jordansissel** [comment]
@guyboertje thank you for entertaining my thoughts about Locale and NumberFormat :)

+1 to move forward as-is.

**Comment by guyboertje** [comment]
Mirrored docs to config comments

**Comment by colinsurprenant** [comment]
LGTM

**Pull request merged**